### PR TITLE
[FLINK-33389][runtime] Support tasks balancing at slot level for Adaptive Scheduler

### PR DESCRIPTION
**Pull request by RocMarshal**
## What is the purpose of the change

[FLINK-33389][runtime] Support tasks balancing at slot level for Adaptive Scheduler

## Brief change log

- Introduce the `SlotSharingStrategy` and related implementations for the original slot sharing strategy and tasks balanced slot sharing strategy.
- Add `CoLocationGroup` related methods for `JobInformation`, `VertexInformation`
- Do some related adaption lines.

## Verifying this change

This change is already covered by existing tests,

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / **don't know**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        return new TaskBalancedExecutionSlotSharingGroupBuilder(
+                        getAllVertices(jobInformation, vertexParallelism),
+                        jobInformation.getSlotSharingGroups(),
+                        jobInformation.getCoLocationGroups())
+                .build().values().stream()
+                        .map(
+                                fromGroup ->
+                                        new ExecutionSlotSharingGroup(
+                                                fromGroup.getExecutionVertexIds()))
+                        .collect(Collectors.toList());
```
Comment (huyuanfeng2018) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java:54]:
```suggestion
            .build().values().stream()
                        .distinct()
                        .map(
                                fromGroup ->
                                        new ExecutionSlotSharingGroup(
                                                fromGroup.getExecutionVertexIds()))
                        .collect(Collectors.toList());
```
missing **distinct**

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    static Map<JobVertexID, List<ExecutionVertexID>> getAllVertices(
+            JobInformation jobInformation, VertexParallelism vertexParallelism) {
+        final Map<JobVertexID, List<ExecutionVertexID>> jobVertexToExecutionVertices =
+                new HashMap<>();
```
Comment (huyuanfeng2018) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java:60]: It seems that **TreeMap** is better

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java (hunk 1 above)
Comment (RocMarshal) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java]: thx for the catching~. 👍 
I'll add a related test case for it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java (hunk 2 above)
Comment (RocMarshal) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java:61]: Based on strategy defined in the https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=263429715#FLIP370:SupportBalancedTasksScheduling-2.1.1InaSlotSharingGroup

In the current slot dimension balancing strategy, we focus on balancing the number of tasks and do not care about the relationship between task predecessor and successor. Therefore, in the subsequent processing, we do not need any information related to the order of tasks temporarily.
In other words, I don't disagree with using TreeMap, we can do it anytime as needed. If the understanding is incorrect, please correct me.  Pls let me know what's your opinon~
Thank you!

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java (hunk 1 above)
Comment (RocMarshal) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/TaskBalancedSlotSharingStrategy.java]: Add the test cases in TaskBalancedSlotSharingStrategyTest

**Comment by RocMarshal** [comment]
blocked by https://github.com/apache/flink/pull/25647

**Comment by RocMarshal** [comment]
Hi, @1996fanrui @huyuanfeng2018  Could you help take a look if you had the free time ? Many thanks !

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/AbstractSlotSharingStrategy.java [hunk 3]
```diff
@@ -95,15 +96,15 @@ public void notifySchedulingTopologyUpdated(
      * topologically sorted.
      */
     @Nonnull
-    static LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> getExecutionVertices(
-            SchedulingTopology topology) {
-        final LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> vertices =
-                new LinkedHashMap<>();
+    static <T> LinkedHashMap<JobVertexID, List<T>> getExecutionVertices(
```
Comment (huyuanfeng2018) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/AbstractSlotSharingStrategy.java:99]: Introducing a function parameter makes it difficult to read. Can add params comments for easier reading?

**Review by huyuanfeng2018 — APPROVED** [review]
overall LGTM.

**Review by huyuanfeng2018 — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedExecutionSlotSharingGroupBuilder.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Execute slot sharing groups builder based on the balanced scheduling strategy. */
+public class TaskBalancedExecutionSlotSharingGroupBuilder {
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedExecutionSlotSharingGroupBuilder.java:46]: Good proposal 👍 and many thanks ! Updated~

**Comment by RocMarshal** [comment]
@flinkbot run azure

**Comment by RocMarshal** [comment]
Hi, @ferenc-csaky Could you help take a look at this ?
Any input is appreciated~
